// @netlify/blobs/dist/main.js
var HTTPMethod;
(function (HTTPMethod) {
    HTTPMethod["Delete"] = "delete";
    HTTPMethod["Get"] = "get";
    HTTPMethod["Put"] = "put";
})(HTTPMethod || (HTTPMethod = {}));
var ResponseType;
(function (ResponseType) {
    ResponseType["ArrayBuffer"] = "arrayBuffer";
    ResponseType["Blob"] = "blob";
    ResponseType["JSON"] = "json";
    ResponseType["Stream"] = "stream";
    ResponseType["Text"] = "text";
})(ResponseType || (ResponseType = {}));
export class Blobs {
    constructor({ authentication, context, fetcher, siteID }) {
        this.context = context ?? 'production';
        this.fetcher = fetcher ?? globalThis.fetch;
        this.siteID = siteID;
        if ('contextURL' in authentication) {
            this.authentication = authentication;
        }
        else {
            this.authentication = {
                apiURL: authentication.apiURL ?? 'https://api.netlify.com',
                token: authentication.token,
            };
        }
        if (fetcher) {
            this.fetcher = fetcher;
        }
        else if (globalThis.fetch) {
            this.fetcher = globalThis.fetch;
        }
        else {
            throw new Error('You must specify a fetch-compatible `fetcher` parameter when `fetch` is not available globally');
        }
    }
    async getFinalRequest(key, method) {
        const finalMethod = method;
        if ('contextURL' in this.authentication) {
            return {
                headers: {
                    authorization: `Bearer ${this.authentication.token}`,
                },
                method: finalMethod,
                url: `${this.authentication.contextURL}/${this.siteID}/${this.context}/${key}`,
            };
        }
        const apiURL = `${this.authentication.apiURL}/api/v1/sites/${this.siteID}/blobs/${key}?context=${this.context}`;
        const headers = { authorization: `Bearer ${this.authentication.token}` };
        const res = await this.fetcher(apiURL, { headers, method });
        const { url } = await res.json();
        return {
            method: finalMethod,
            url,
        };
    }
    async makeStoreRequest(key, method, extraHeaders, body) {
        const { headers: baseHeaders = {}, method: finalMethod, url } = await this.getFinalRequest(key, method);
        const headers = {
            ...baseHeaders,
            ...extraHeaders,
        };
        if (method === HTTPMethod.Put) {
            headers['cache-control'] = 'max-age=0, stale-while-revalidate=60';
        }
        return await this.fetcher(url, { body, headers, method: finalMethod });
    }
    async delete(key) {
        return await this.makeStoreRequest(key, HTTPMethod.Delete);
    }
    async get(key, options) {
        const { type } = options ?? {};
        const res = await this.makeStoreRequest(key, HTTPMethod.Get);
        if (res.status === 404) {
            return null;
        }
        if (type === undefined || type === ResponseType.Text) {
            return res.text();
        }
        if (type === ResponseType.ArrayBuffer) {
            return res.arrayBuffer();
        }
        if (type === ResponseType.Blob) {
            return res.blob();
        }
        if (type === ResponseType.JSON) {
            return res.json();
        }
        if (type === ResponseType.Stream) {
            return res.body;
        }
        throw new Error(`Invalid 'type' property: ${type}. Expected: arrayBuffer, blob, json, stream, or text.`);
    }
    async set(key, data) {
        await this.makeStoreRequest(key, HTTPMethod.Put, {}, data);
    }
    async setJSON(key, data) {
        const payload = JSON.stringify(data);
        const headers = {
            'content-type': 'application/json',
        };
        await this.makeStoreRequest(key, HTTPMethod.Put, headers, payload);
    }
}
